VS: replace IServiceProvider with IAsyncServiceProvider

zhili1208 opened the pull request:
https://github.com/NuGet/Home/issues/6771

jainaashish commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreOperationLogger.cs:
```diff
@@ -464,7 +471,7 @@ private int GetMSBuildOutputVerbositySetting()
             ThreadHelper.ThrowIfNotOnUIThread();
 
-            var dte = _serviceProvider.GetDTE();
+            var dte = _taskFactory.Run(async () => await _asyncServiceProvider.GetDTEAsync());
```
This makes additional dependency for threadpool thread to get the dte instance even though its already on UI thread which will create problems when threadpool is already exhausted. So since this is a private api make it async or if you can't then just use `GetDte`.

jainaashish commented on src/NuGet.Clients/NuGet.Tools/Commands/ShowErrorsCommand.cs:
```diff
@@ -20,28 +21,26 @@ internal sealed class ShowErrorsCommand : ICommand
+                throw new ArgumentNullException(nameof(asyncServiceProvider));
             }
             // get all services we need for display and activation of the NuGet output pane
             _vsOutputWindow = new Lazy<IVsOutputWindow>(() =>
```
convert it to `AsyncLazy`

jainaashish commented on src/NuGet.Clients/NuGet.Tools/Commands/ShowErrorsCommand.cs:
```diff
@@ -20,28 +21,26 @@ internal sealed class ShowErrorsCommand : ICommand
                 });
             });
 
             _vsUiShell = new Lazy<IVsUIShell>(() =>
```
same as above, convert it to `AsyncLazy`

jainaashish commented on src/NuGet.Clients/NuGet.Tools/OutputConsoleLogger.cs:
```diff
@@ -51,9 +56,7 @@ public class OutputConsoleLogger : INuGetUILogger, IDisposable
 
             NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
```
this is problematic to try to switch to UI thread in MEF constructor. Please make `_dte` as `AsyncLazy` or add a separate `initializeAsync` api

jainaashish reviewed (changes requested):
Few suggestions to improve it further, but overall looks good.

zhili1208 commented:
addressed most feedback, talked to @mishra14 only thing I can't change is can't remove "switch to UI thread" in OutputConsoleLogger because we need to hook up some solution event there.

jainaashish commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs:
```diff
@@ -32,31 +31,53 @@ namespace NuGet.VisualStudio
             _vsProjectAdapterProvider = vsProjectAdapterProvider ?? throw new ArgumentNullException(nameof(vsProjectAdapterProvider));
             _getLockFileOrNull = BuildIntegratedProjectUtility.GetLockFileOrNull;
+
+            _dte = new Microsoft.VisualStudio.Threading.AsyncLazy<EnvDTE.DTE>(
```
rebase from dev, this has changed.

jainaashish reviewed (approved):
Looks good to me. Run it through RPS and test it thoroughly from UI and powershell console. May be ask vendor to do that.

nkolev92 commented:
@zhili1208 
Can you please create a matching issue for this? Or link if there already is one. 

Also can you please add more details/pointers to make reviewing easier?

nkolev92 commented on src/NuGet.Clients/NuGet.Console/PowerConsole/PowerConsoleWindow.cs:
```diff
@@ -23,7 +23,6 @@ internal class PowerConsoleWindow : IPowerConsoleWindow, IHostInitializer, IDisp
         private HostInfo _activeHostInfo;
 
         [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
-        [Import(typeof(SVsServiceProvider))]
```
>ServiceProvider [](start = 26, length = 15)

How is this populated now? 
Should we use the async provider here?

zhili1208 replied:
I can't change this, because this not called on async path, we have some property which be called on sync path.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs:
```diff
@@ -32,26 +31,37 @@ namespace NuGet.VisualStudio
             Lazy<ISettings> settings,
             Lazy<IVsSolutionManager> solutionManager,
             [Import("VisualStudioActivityLogger")]
             Lazy<NuGet.Common.ILogger> logger,
```
>NuGet [](start = 17, length = 5)

I would still keep the logger as the last parameter.

nkolev92 reviewed (commented):
Looks good overall, but I'm curious why the PowerConsoleWindow is not updated, and a style suggestion.

zhili1208 committed:
VS: replace IServiceProvider with IAsyncServiceProvider
